### ISSUE TITLE
Crash when testing VLCJ 3.x in Linux with JDK 14 Early Access (jpackage)

### ISSUE THREAD

**Issue by motohead99** [caprica/vlcj#897]
I'm using VLCJ 3.12.1 and an Open JDK that has "14-jpackage" as the version.

You can find the Open JDK 14 Early Access with jpackage here:
https://jdk.java.net/jpackage/

For Linux specifically I'm using this one:
https://download.java.net/java/early_access/jpackage/1/openjdk-14-jpackage+1-35_linux-x64_bin.tar.gz

In a nutshell, the jpackage tool lets me build a Java based application with a native executable, a "minified built-in JRE", and an installer to install the application. It's not yet part of the mainstream Java as it still has a few bugs to iron out first.

In Windows, VLCJ continues to work great with the JDK 14 Early Access with jpackage. However, I'm now testing in Linux and it fails to initialize due to an error in Version.java. Reviewing the code on github shows that the specific code in question is applying a fix for JDK >1.7 on Linux only (which is presumably why I don't see it on Windows).

Version.java (exception from here):
https://github.com/caprica/vlcj/blob/3f78f007a4e10a34b0c5a0a10cad559ec7cd933d/src/main/java/uk/co/caprica/vlcj/version/Version.java#L85

MediaPlayerFactory.java (Linux JDK >1.7 code):
https://github.com/caprica/vlcj/blob/3f78f007a4e10a34b0c5a0a10cad559ec7cd933d/src/main/java/uk/co/caprica/vlcj/player/MediaPlayerFactory.java#L150

I'm guessing:
- VERSION_PATTERN needs to be more flexible, for example, (\\d+).*\\. ... to allow for characters after the first number, or
- modify MediaPlayerFactory's getJavaVersion to parse the leading number from the front of System.getProperty("java.version") and discard the "-jpackage", then append the ".0.0"

The exception is:
`java.lang.RuntimeException: Exception in Application start method
    at javafx.graphics/com.sun.javafx.application.LauncherImpl.launchApplication1(LauncherImpl.java:900)
    at javafx.graphics/com.sun.javafx.application.LauncherImpl.lambda$launchApplication$2(LauncherImpl.java:195)
    at java.base/java.lang.Thread.run(Thread.java:830)`

 --- snip ---
`Caused by: java.lang.IllegalArgumentException: Can't parse version from '14-jpackage.0.0'
    at vlcj/uk.co.caprica.vlcj.version.Version.<init>(Version.java:85)
    at vlcj/uk.co.caprica.vlcj.player.MediaPlayerFactory.<clinit>(MediaPlayerFactory.java:150)
    ... 16 more`

**Comment by motohead99** [comment]
Sorry, I hardly had any time to look into this today. I was able to convert my code to use the vlcj 4.x API so that it builds but I was in a rush and didn't check things thoroughly. The main thing that might be lacking is the locking mechanism that was there with 3.x. Maybe it's just no longer needed (i.e. mediaPlayer.lock() and mediaPlayer.unlock() in the RenderCallback). I don't see any lock/unlock methods in the new API so I'm guessing that they're no longer needed.

Again, this is all really rushed.

With vlcj 4.2.0 my app builds fine in Linux and runs without crashing. I see the first frame of video so fundamentally it should be ok. However, I'm unable to start playback in Linux for some reason (but in Windows it works fine). I suspect that this is a bug in my code and not vlcj. So I'll need to look into this further next week when I have more time for it. Maybe I missed something when porting to the new API.

So in a nutshell, this issue is probably fixed by 4.2.0. I'm not sure if you'd want to fix it for 3.x if it's no longer maintained.

Hopefully I'll have a more solid/stable answer for you next week. Hope you have a good weekend.

**Comment by caprica** [comment]
Just FYI on "callback" media players, as you've found now these replace "direct" media players - they should be much easier to implement and actually perform a little better in vlcj 4.x+ when compared with vlcj 3.x.

There is indeed no need for explicit lock/unlock anymore. I don't recall the fine detail of the implementation but I think the video buffer is copied before the native callback returns so there's no chance of any partial buffers or corrupted buffers. Even if that vague explanation is not quite right, the one thing I can say is that I use these callback media players a LOT with vlcj-4.x and I never saw any problems with the current implementation.

I use Linux for 99% of my development activities so I know for sure it will work there, hopefully you will find your problem easily.

Thanks a lot for the feedback.

**Comment by motohead99** [comment]
Hey, sorry for the late response. Migrating to vlcj 4.x fixes this issue so you can close it if you see fit.
I did have an issue getting vlcj 4.x to work in Mac, due to this whole JPMS and the use of com.apple.eawt. I'm going to take another stab at that though. For now, I just built vlcj 4.x myself and removed the com.apple.eawt dependencies since I don't need full screen playback anyways. I have a feeling that I could get around this though by not using the vlcj-osx-stubs and passing in runtime arguments to the JVM to allow the use of internal Java APIs. I'll give that a shot next.
This specific issue is fixed with vlcj 4.x though.

**Comment by caprica** [comment]
OK, I'll close it because I don't really want to make a new release in the 3.x series.
